bigquery/tests/unit/test_job.py

```diff
@@ -4008,21 +4008,37 @@ def test_estimated_bytes_processed(self):
         self.assertEqual(job.estimated_bytes_processed, est_bytes)
 
     def test_result(self):
+        from google.cloud.bigquery.table import RowIterator
+
         query_resource = {
             "jobComplete": True,
             "jobReference": {"projectId": self.PROJECT, "jobId": self.JOB_ID},
             "schema": {"fields": [{"name": "col1", "type": "STRING"}]},
+            "totalRows": "2",
         }
-        connection = _make_connection(query_resource, query_resource)
+        tabledata_resource = {
+            "totalRows": "1",
+            "pageToken": None,
+            "rows": [{"f": [{"v": "abc"}]}],
+        }
+        connection = _make_connection(query_resource, tabledata_resource)
         client = _make_client(self.PROJECT, connection=connection)
         resource = self._make_resource(ended=True)
         job = self._get_target_class().from_api_repr(resource, client)
 
         result = job.result()
 
-        self.assertEqual(list(result), [])
+        self.assertIsInstance(result, RowIterator)
+        self.assertEqual(result.total_rows, 2)
+
+        rows = list(result)
+        self.assertEqual(len(rows), 1)
+        self.assertEqual(rows[0].col1, "abc")
+        self.assertEqual(result.total_rows, 1)
```

Comment:
Do you mean to test explicitly the case where the initial query result resource has a different number of rows than the count got by iteration?  If so, then a comment above stating that would be helpful.

Reply:
Yes, that's intentional. You're right that it looks wrong. I suspect it'll be quite rare in practice, but  it can happen, especially when append query jobs are involved. Added comments in 99441d71ea1233b8bd8089c06d835298b9cff84d.